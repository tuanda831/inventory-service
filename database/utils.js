exports.TRUNCATE_ALL_TABLES = `
DO $$
DECLARE
    tables CURSOR FOR
      SELECT    TABLENAME
      FROM      PG_TABLES
      WHERE     SCHEMANAME = 'public'
      AND       TABLENAME NOT LIKE 'knex_%'
      ORDER BY  TABLENAME;
BEGIN
    FOR tblRec IN tables LOOP
      EXECUTE 'TRUNCATE TABLE ' || tblRec.tablename || ' CASCADE;';
    END LOOP;
END$$;
`;