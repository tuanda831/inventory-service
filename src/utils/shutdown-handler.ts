import { LoggerInterface } from '../dal/logger/logger.providers';

const signalsNames: NodeJS.Signals[] = ['SIGTERM', 'SIGINT', 'SIGHUP'];

export const shutdownHandler = (logger: LoggerInterface) => {
  signalsNames.forEach((signalName) =>
    process.on(signalName, (signal) => {
      logger.log(`Retrieved signal: ${signal}, application terminated`);
      process.exit(0);
    }),
  );

  process.on('uncaughtException', (error: Error) => {
    logger.error({ err: error });
    process.exit(1);
  });

  process.on('unhandledRejection', (reason, promise) => {
    logger.error(`Unhandled Promise Rejection, reason: ${reason}`);
    promise.catch((err: Error) => {
      logger.error({ err });
      process.exit(1);
    });
  });
};
