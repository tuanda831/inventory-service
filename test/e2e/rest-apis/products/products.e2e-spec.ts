import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { ApisModule } from '../../../../src/rest-apis/apis.module';
import { ProductCreationRequest } from 'src/rest-apis/requests/products/product-creation.request';
import { TRUNCATE_ALL_TABLES } from '../../utils';
import { DataSource } from 'typeorm';

describe('ApiController (e2e)', () => {
  let app: INestApplication;
  let dbClient: DataSource;

  beforeAll(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [ApisModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();

    app.enableShutdownHooks();

    dbClient = app.get(DataSource);
    await dbClient.query(TRUNCATE_ALL_TABLES);
  });

  afterAll(async () => {
    await app.close();
  });

  describe('Products', () => {
    describe('Get empty products', () => {
      it('should get empty list', async () => {
        const res = await request(app.getHttpServer()).get('/products');
        expect(res.status).toBe(200);
        expect(res.body).toMatchSnapshot();
      });
    });

    describe('Create product', () => {
      const productReq: ProductCreationRequest = {
        sku: '123',
        barcode: '99999',
        name: 'New Product',
        description: 'Product Description here',
        images: ['https://test1.link', 'https://test2.link'],
        price: 1.98,
      };

      it('should create a product', async () => {
        const res = await request(app.getHttpServer())
          .post('/products')
          .send(productReq);

        expect(res.status).toBe(201);
        expect(res.body).toMatchSnapshot({
          data: {
            id: expect.any(String),
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
          },
        });
      });
    });

    describe('Get products', () => {
      it('should get a products', async () => {
        const res = await request(app.getHttpServer()).get('/products');
        expect(res.status).toBe(200);
        expect(res.body).toMatchSnapshot({
          data: [
            {
              id: expect.any(String),
              createdAt: expect.any(String),
              updatedAt: expect.any(String),
            },
          ],
        });
      });
    });
  });
});
