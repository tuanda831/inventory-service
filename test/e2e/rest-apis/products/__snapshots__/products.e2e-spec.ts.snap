// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApiController (e2e) Products Create product should create a product 1`] = `
Object {
  "data": Object {
    "barcode": "99999",
    "createdAt": Any<String>,
    "deletedAt": null,
    "description": "Product Description here",
    "id": Any<String>,
    "images": Array [
      "https://test1.link",
      "https://test2.link",
    ],
    "name": "New Product",
    "price": 1.98,
    "restoredAt": null,
    "shortCode": "New-Product-123",
    "sku": "123",
    "updatedAt": Any<String>,
  },
  "message": "\\"Product\\" is Created!",
}
`;

exports[`ApiController (e2e) Products Get empty products should get empty list 1`] = `
Object {
  "data": Array [],
  "meta": Object {
    "limit": 0,
    "offset": 0,
    "totalCount": 0,
  },
}
`;

exports[`ApiController (e2e) Products Get products should get a products 1`] = `
Object {
  "data": Array [
    Object {
      "barcode": "99999",
      "createdAt": Any<String>,
      "deletedAt": null,
      "description": "Product Description here",
      "id": Any<String>,
      "images": Array [
        "https://test1.link",
        "https://test2.link",
      ],
      "name": "New Product",
      "price": "1.98",
      "restoredAt": null,
      "shortCode": "New-Product-123",
      "sku": "123",
      "updatedAt": Any<String>,
    },
  ],
  "meta": Object {
    "limit": 0,
    "offset": 0,
    "totalCount": 0,
  },
}
`;
